File: typescript/packages/core/src/cron-agent-surface.ts
import type { AgentRuntime } from './runtime.js';
import { CronScheduler, type CronJob } from './cron-scheduler.js';

export interface CronAgentSurfaceOptions {
  runtime: AgentRuntime;
}

export class CronAgentSurface {
  private readonly scheduler: CronScheduler;

  constructor(private readonly options: CronAgentSurfaceOptions) {
    this.scheduler = new CronScheduler(async (job) => {
      const generator = this.options.runtime.router.route({
        sessionId: job.sessionId,
        input: { role: 'user', content: job.prompt },
      });
      let result = await generator.next();
      while (!result.done) {
        result = await generator.next();
      }
      void result.value;
    });
  }

  register(job: CronJob): void {
    this.scheduler.register(job);
  }

  unregister(jobId: string): void {
    this.scheduler.unregister(jobId);
  }

  stopAll(): void {
    this.scheduler.stopAll();
  }
}

File: python/src/agent_framework/core/cron_agent_surface.py
from __future__ import annotations

from agent_framework.core.cron_scheduler import CronJob, CronScheduler
from agent_framework.core.domain import AgentRunRequest


class CronAgentSurface:
    def __init__(self, runtime) -> None:
        self._runtime = runtime
        self._scheduler = CronScheduler(self._handle_job)

    async def _handle_job(self, job: CronJob) -> None:
        async for _event in self._runtime.router.route(
            AgentRunRequest(session_id=job.session_id, input={"role": "user", "content": job.prompt})
        ):
            pass

    def register(self, job: CronJob) -> None:
        self._scheduler.register(job)

    def unregister(self, job_id: str) -> None:
        self._scheduler.unregister(job_id)

    def stop_all(self) -> None:
        for job_id in list(self._scheduler._tasks.keys()):
            self._scheduler.unregister(job_id)
